crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -625,4 +650,18 @@ mod tests {
         let _ = world.run_system(nested_id);
         assert_eq!(*world.resource::<Counter>(), Counter(5));
     }
+
+    #[test]
+    fn system_with_input_ref() {
+        fn with_ref(InRef(input): InRef<u8>, mut counter: ResMut<Counter>) {
+            counter.0 += *input;
+        }
+
+        let mut world = World::new();
+        world.insert_resource(Counter(0));
+
+        let id = world.register_system(with_ref);
+        world.run_system_with_input(id, &2).unwrap();
+        assert_eq!(*world.resource::<Counter>(), Counter(2));
+    }
```

Comment:
Might be worth including a test using the mutable reference type as well, since I think that better showcases the most interesting functionality of this change. Something like:

```rust
fn get_count(InMut(count): InMut<usize>, query: Query<Animals>) {
    *count = query.iter().count();
}
```